hail/src/main/scala/is/hail/expr/ir/EmitStreamDistribute.scala

```diff
@@ -0,0 +1,303 @@
+package is.hail.expr.ir
+
+import is.hail.annotations.Region
+import is.hail.asm4s.{Code, Value, const}
+import is.hail.expr.ir.functions.MathFunctions
+import is.hail.expr.ir.orderings.StructOrdering
+import is.hail.expr.ir.streams.EmitStream
+import is.hail.io.OutputBuffer
+import is.hail.types.VirtualTypeWithReq
+import is.hail.types.physical.{PBooleanRequired, PCanonicalArray, PCanonicalInterval, PCanonicalStringRequired, PCanonicalStruct, PContainer, PInt32Required}
+import is.hail.types.physical.stypes.EmitType
+import is.hail.types.physical.stypes.concrete.{SIndexablePointer, SIndexablePointerCode, SJavaString, SStackStruct, SStackStructCode}
+import is.hail.types.physical.stypes.interfaces.{SBaseStruct, SIndexableCode, SIndexableValue, SStreamCode, SStringCode, primitive}
+import is.hail.types.physical.stypes.primitives.{SBooleanCode, SInt32, SInt32Code}
+import is.hail.types.virtual.{TArray, TStruct}
+import is.hail.utils._
+import is.hail.asm4s._
+
+object EmitStreamDistribute {
+
+  def emit(emitter: Emit[_], ir: StreamDistribute, cb: EmitCodeBuilder, region: Value[Region], env: EmitEnv, typeWithReqx: IR => VirtualTypeWithReq, container: Option[AggContainer], loopEnv: Option[Env[LoopRef]]): IEmitCode = {
```

Comment:
Splitting the emitter into a bunch of mutually recursive functions distributed across a bunch of files adds a lot of unnecessary complexity. I actually want to stop doing that in `EmitStream` and `EmitNDArray`. But here it's easy to avoid: just make the function in a separate file take inputs at the level of `Code/SCode`, not `IR`. E.g. in a separate file:
```scala
  def streamDistribute(
    cb: EmitCodeBuilder,
    requestedSplittersAndEndsVal: SIndexableValue,
    childStream: SStreamCode,
    pathVal: SValue,
    spec: AbstractTypedCodecSpec,
    resultType: TStruct,
    region: Value[Region]
  ): SIndexablePointerCode = {
    val mb = cb.emb
    val pivotsPType = requestedSplittersAndEndsVal.st.storageType().asInstanceOf[PCanonicalArray]
    ...
  }
```
and in Emit:
```scala
      case ir@StreamDistribute(child, pivots, path, spec) =>
        val resultType = ir.typ.asInstanceOf[TArray].elementType.asInstanceOf[TStruct]
        emitI(path).flatMap(cb) { case pathCode: SStringCode =>
          val pathVal = pathCode.memoize(cb, "stream_dist_output_path")

          emitI(pivots).flatMap(cb) { case pivotsCode: SIndexableCode =>
            emitStream(child, cb, region).map(cb) { case childStream: SStreamCode =>
              val requestedSplittersAndEndsVal = pivotsCode.memoize(cb, "stream_dist_pivots")
              EmitStreamDistribute.streamDistribute(cb, requestedSplittersAndEndsVal, childStream, pathVal, spec, resultType, region)
            }
          }
        }
```